### PR TITLE
migrate meta's image attention to HF equivalent

### LINKED ISSUE #30086: `test_llama_image_attention.py`
replace llama_models import with hf import is the end goal.

https://github.com/tenstorrent/tt-metal/blob/bf338f0844e710e58de614b919f4e959b6f133a1/models/tt_transformers/tests/multimodal/test_llama_image_attention.py#L50

### PR DESCRIPTION
**Pull request by epam-ioannis-alexiou**
### Ticket
Close https://github.com/tenstorrent/tt-metal/issues/30086

### Problem description
Image attention for the vision branch of mulitmodal Llama needs to be migrated from meta lib to HF.

### What's changed
The unmapped HF weights are used for the HF model whereas TT model uses remapped HF weights. A function to load and extract the revelavant weights is used so the  weights can be allocated to the reference layer. Also HF build the attention maps internally in this example the attention maps are explicitly defined in the script and passed to the attention blocks.

**Note:**
HF internally may use different attention implementations, by default eager mode is always active. This mode uses an attention implementation that if combined with making a continuous block of memory using `.contiguous()` and `nn.Linear` layer generates minor numerical deviations in the generated tensor. Thus when testing Llama model attention its implementation is switched to the following 
"config.vision_config._attn_implementation = "sdpa"" to use `torch.nn.functional.scaled_dot_product_attention` as in meta lib (this diminishes any numerical fluctuations).

### Checklist
- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/20128895035) CI passed  for `test_llama_image_attention.py` with relevant CI tasks `t3k llama3.2-90b-vision tests`, `t3k n300 mesh llama3.2-11b-vision tests`, `t3k llama3.2-11b-vision tests`

### CODE REVIEW

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 1]
```diff
@@ -63,12 +108,11 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds, ensure
 
     # Create PT input
     ar = torch.tensor([[1, 2]])
-    pt_block_input = (torch.rand(batch, num_chunks, ntok, dim) * 2) - 1
+    pt_block_input = ((torch.rand(batch, num_chunks, ntok, dim) * 2) - 1) * 0.1
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: @epam-ioannis-alexiou @maksim-tsishkouski-epam and I had a discussion on Slack about this. There was a small drop in PCC without the `* 0.1` change. Ioannis is going to investigate a little further into this -- the first step is to pull in changes on `main` as Maksim's PR removed LLAMA_DIR, which means the currently passing test on `main` is already loading HF checkpoints. The new HF reference model loads the same checkpoints. Our expectation is that there could some dtype mismatch between the current reference model and the new reference model.

Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: I re-run CI tests without * 0.1 which was a left over before I switched to functional sdpa implementation. Now it passes higher than 0.999 pcc. I have also included HF padding in the code and removed llama padding implementation. 
The CI result can be found in the description and also here [T3K unit-tests](https://github.com/tenstorrent/tt-metal/actions/runs/20061476433).

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 2]
```diff
@@ -104,9 +154,9 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds, ensure
     tt_out = ttnn.slice(tt_out, (0, 0, 0, 0), (batch, num_chunks, ntok, dim))
     tt_output_torch = ttnn.to_torch(tt_out, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=0))[0, :, :, :]
 
-    reference_output = reference_model(pt_block_input, mask=mask)
-    reference_output = reference_output.reshape(batch, num_chunks, ntok + npad, dim)
-    reference_output = encoder_utils.contract_num_tokens_from_mult8(reference_output, npad)
+    reference_output = reference_model(pt_block_input, attention_mask=mask)[0]
+    reference_output = reference_output.reshape(batch, num_chunks, ntok - slice_index, dim)
```
Comment (maksim-tsishkouski-epam) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py:143]: Wouldn't it fail if slice_index is None (if num_padding_patches==0)?

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 3]
```diff
@@ -17,6 +18,47 @@
...
+    return partial_state_dict
+
+
+def build_encoder_attention_mask(
+    x: torch.Tensor,
+    ar: torch.Tensor,
+    ntok: int,
+    num_chunks: int,
+    n_heads: int,
+):
```
Comment (maksim-tsishkouski-epam) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: Seems like the function is used in multiple places (in other tests as well), maybe makes sense to move it to some "common" file, e.g. `models/common/llama_models.py`

**Review by maksim-tsishkouski-epam — CHANGES_REQUESTED** [review]

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 4]
```diff
@@ -40,14 +82,18 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds, ensure
...
-    heads = model_args.vision_attn_n_heads
     ntok = model_args.vision_chunk_ntok
-    reference_model = llama_reference_mod.ImageAttention(dim=dim, head_dim=dim // heads, n_heads=heads)
+
+    # config contains paramters for the whole multimodal network the subeset of vision branch is chosen instead
+    config = AutoConfig.from_pretrained(os.getenv("HF_MODEL"))
+    config.vision_config._attn_implementation = "sdpa"
+    reference_model = MllamaVisionAttention(config.vision_config)
+    # partial loading of HF safetensors to match model graph expected dimensionality of the loaded weights
+    partial_state_dict = load_partial_weights(
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: nit: `os.getenv("HF_MODEL")` is used a second time here; it would be nicer to refactor this code to `model_name_or_path = os.getenv("HF_MODEL")` before the first use of it in line 90.

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 5]
```diff
@@ -65,10 +111,14 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds, ensure
     ar = torch.tensor([[1, 2]])
     pt_block_input = (torch.rand(batch, num_chunks, ntok, dim) * 2) - 1
     tt_attention_input = pt_block_input.clone()
-    # Do PT padding
-    pt_block_input, npad = encoder_utils.expand_num_tokens_to_mult8(pt_block_input)
+    num_padding_patches = (8 - (pt_block_input.shape[-2] % 8)) % 8
+    # Compute padding tuple for pad function
+    padding = (0, 0, 0, num_padding_patches)  # (pad_left, pad_right, pad_left for dim -2, pad_right for dim -2)
+    # Pad the tensor
+    pt_block_input = F.pad(pt_block_input, padding, mode="constant", value=0)
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: 8 looks like a magic number here and it can hurt readability. Maybe a good idea to refactor this code a function that takes a descriptive name like the Meta one -- expand_num_tokens_to_mult8?

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py [hunk 6]
```diff
@@ -65,10 +111,14 @@ def test_attention_inference(batch, num_chunks, mesh_device, reset_seeds, ensure
...
+    slice_index = -num_padding_patches if num_padding_patches > 0 else None
```
Comment (gwangTT) [inline models/tt_transformers/tests/multimodal/test_llama_image_attention.py]:
```suggestion
    slice_index = -num_padding_patches if num_padding_patches > 0 else pt_block_input.shape[-2]
```
would be better? Like Maksim commented, None would error out later in the code.

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 2 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py:143]: good point, I have added a wrapper function in this commit to deal with that [06e9523](https://github.com/tenstorrent/tt-metal/pull/33553/commits/06e9523bac809753861c20b244e269ddb897c29e)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 3 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: Yes this is the main attention mask builder used in most models that have attention layers. I have moved it into `models/tt_transformers/tt/common.py`  found in [06e9523](https://github.com/tenstorrent/tt-metal/pull/33553/commits/06e9523bac809753861c20b244e269ddb897c29e)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 4 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: no prb I have replaced these with a single string variable in this [06e9523](https://github.com/tenstorrent/tt-metal/pull/33553/commits/06e9523bac809753861c20b244e269ddb897c29e)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 5 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: I have added 2 functions to deal with padding and padding removal in this commit that follow the logic of the comment above [06e9523](https://github.com/tenstorrent/tt-metal/pull/33553/commits/06e9523bac809753861c20b244e269ddb897c29e)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 6 above)
Comment (epam-ioannis-alexiou) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: it is refactored into wrapper functions in this [06e9523](https://github.com/tenstorrent/tt-metal/pull/33553/commits/06e9523bac809753861c20b244e269ddb897c29e)

**Review by maksim-tsishkouski-epam — APPROVED** [review]

**Review by gwangTT — APPROVED** [review]
LGTM. Thanks, Ioannis! The code looks clean and readable.

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 2 above)
Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py:143]: It is this commit that addressed this comment: [0adfcb9](https://github.com/tenstorrent/tt-metal/pull/33553/commits/0adfcb9b278a8e9a65b6f4ac8da4559ba0d6a099)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 3 above)
Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: This commit is the addressor of the comment: [0adfcb9](https://github.com/tenstorrent/tt-metal/pull/33553/commits/0adfcb9b278a8e9a65b6f4ac8da4559ba0d6a099)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 4 above)
Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: This commit is the addressor of the comment: [0adfcb9](https://github.com/tenstorrent/tt-metal/pull/33553/commits/0adfcb9b278a8e9a65b6f4ac8da4559ba0d6a099)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 5 above)
Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: This commit is the addressor of the comment: [0adfcb9](https://github.com/tenstorrent/tt-metal/pull/33553/commits/0adfcb9b278a8e9a65b6f4ac8da4559ba0d6a099)

File: models/tt_transformers/tests/multimodal/test_llama_image_attention.py (hunk 6 above)
Comment (gwangTT) [reply models/tt_transformers/tests/multimodal/test_llama_image_attention.py]: This commit is the addressor of the comment: [0adfcb9](https://github.com/tenstorrent/tt-metal/pull/33553/commits/0adfcb9b278a8e9a65b6f4ac8da4559ba0d6a099)
